### PR TITLE
Added initial support for OCP-TV output format

### PR DESCRIPTION
**Pull request by sksekar**
This is to add initial support. Yet to add support for validators, pass/fail etc.

Sample OCP-TV output:

```
    {"schemaVersion": {"major": 2, "minor": 0}, "sequenceNumber": 0, "timestamp": "2024-08-19T00:30:31.094657Z"}
    {"testRunArtifact": {"testRunStart": {"name": "pci_lmt", "version": "1.3.1", "commandLine": "-o ocp -c lmt.cfg auto", "parameters": {}, "dutInfo": {"dutInfoId": "319706019", "name": "hostname", "platformInfos": [], "softwareInfos": [], "hardwareInfos": []}}}, "sequenceNumber": 1, "timestamp": "2024-08-19T00:30:31.094946Z"}
    {"testStepArtifact": {"testStepId": "0", "testStepStart": {"name": "Rcvr:6 Step:5 Ann:CPU Tx -> Retimer Rx (Timing left)"}}, "sequenceNumber": 2, "timestamp": "2024-08-19T00:30:31.095273Z"}
    {"testStepArtifact": {"testStepId": "0", "measurement": {"name": "BDF:0000:1a:00.0 Lane:0", "value": 0.0, "validators": []}}, "sequenceNumber": 3, "timestamp": "2024-08-19T00:30:36.691321Z"}
    {"testStepArtifact": {"testStepId": "0", "measurement": {"name": "BDF:0000:1a:00.0 Lane:1", "value": 0.0, "validators": []}}, "sequenceNumber": 4, "timestamp": "2024-08-19T00:30:36.691972Z"}
    {"testStepArtifact": {"testStepId": "0", "measurement": {"name": "BDF:0000:1a:00.0 Lane:2", "value": 0.0, "validators": []}}, "sequenceNumber": 5, "timestamp": "2024-08-19T00:30:36.692493Z"}
    {"testStepArtifact": {"testStepId": "0", "measurement": {"name": "BDF:0000:1a:00.0 Lane:3", "value": 0.0, "validators": []}}, "sequenceNumber": 6, "timestamp": "2024-08-19T00:30:36.692845Z"}
    {"testStepArtifact": {"testStepId": "0", "measurement": {"name": "BDF:0000:1a:00.0 Lane:4", "value": 0.0, "validators": []}}, "sequenceNumber": 7, "timestamp": "2024-08-19T00:30:36.693209Z"}
    {"testStepArtifact": {"testStepId": "0", "measurement": {"name": "BDF:0000:1a:00.0 Lane:5", "value": 0.0, "validators": []}}, "sequenceNumber": 8, "timestamp": "2024-08-19T00:30:36.693568Z"}
    {"testStepArtifact": {"testStepId": "0", "measurement": {"name": "BDF:0000:1a:00.0 Lane:6", "value": 0.0, "validators": []}}, "sequenceNumber": 9, "timestamp": "2024-08-19T00:30:36.693947Z"}
    {"testStepArtifact": {"testStepId": "0", "measurement": {"name": "BDF:0000:1a:00.0 Lane:7", "value": 0.0, "validators": []}}, "sequenceNumber": 10, "timestamp": "2024-08-19T00:30:36.694338Z"}
    {"testStepArtifact": {"testStepId": "0", "testStepEnd": {"status": "COMPLETE"}}, "sequenceNumber": 11, "timestamp": "2024-08-19T00:30:36.694586Z"}
    {"testStepArtifact": {"testStepId": "1", "testStepStart": {"name": "Rcvr:6 Step:6 Ann:CPU Tx -> Retimer Rx (Timing left)"}}, "sequenceNumber": 12, "timestamp": "2024-08-19T00:30:36.694755Z"}
    {"testStepArtifact": {"testStepId": "1", "measurement": {"name": "BDF:0000:1a:00.0 Lane:0", "value": 0.0, "validators": []}}, "sequenceNumber": 13, "timestamp": "2024-08-19T00:30:42.283159Z"}
    {"testStepArtifact": {"testStepId": "1", "measurement": {"name": "BDF:0000:1a:00.0 Lane:1", "value": 0.0, "validators": []}}, "sequenceNumber": 14, "timestamp": "2024-08-19T00:30:42.283600Z"}
    {"testStepArtifact": {"testStepId": "1", "measurement": {"name": "BDF:0000:1a:00.0 Lane:2", "value": 0.0, "validators": []}}, "sequenceNumber": 15, "timestamp": "2024-08-19T00:30:42.283974Z"}
    {"testStepArtifact": {"testStepId": "1", "measurement": {"name": "BDF:0000:1a:00.0 Lane:3", "value": 0.0, "validators": []}}, "sequenceNumber": 16, "timestamp": "2024-08-19T00:30:42.284312Z"}
    {"testStepArtifact": {"testStepId": "1", "measurement": {"name": "BDF:0000:1a:00.0 Lane:4", "value": 0.0, "validators": []}}, "sequenceNumber": 17, "timestamp": "2024-08-19T00:30:42.284669Z"}
    {"testStepArtifact": {"testStepId": "1", "measurement": {"name": "BDF:0000:1a:00.0 Lane:5", "value": 0.0, "validators": []}}, "sequenceNumber": 18, "timestamp": "2024-08-19T00:30:42.285002Z"}
    {"testStepArtifact": {"testStepId": "1", "measurement": {"name": "BDF:0000:1a:00.0 Lane:6", "value": 0.0, "validators": []}}, "sequenceNumber": 19, "timestamp": "2024-08-19T00:30:42.285358Z"}
    {"testStepArtifact": {"testStepId": "1", "measurement": {"name": "BDF:0000:1a:00.0 Lane:7", "value": 0.0, "validators": []}}, "sequenceNumber": 20, "timestamp": "2024-08-19T00:30:42.287100Z"}
    {"testStepArtifact": {"testStepId": "1", "testStepEnd": {"status": "COMPLETE"}}, "sequenceNumber": 21, "timestamp": "2024-08-19T00:30:42.287347Z"}
    {"testRunArtifact": {"testRunEnd": {"status": "COMPLETE", "result": "PASS"}}, "sequenceNumber": 22, "timestamp": "2024-08-19T00:30:42.287470Z"}
```

### CODE REVIEW

File: src/pci_lmt/results.py
Comment (mimir-d) [inline src/pci_lmt/results.py:73]: i strongly suggest that you use a contextmanager here; see `run.scope() / test.scope()` in ocptv; this way, other code in pci_lmt wont forget to call the end methods, for either runs or steps

File: src/pci_lmt/results.py
Comment (mimir-d) [inline src/pci_lmt/results.py:107]: typo "oct", and a simpler name works "OcptvReporter"

File: src/pci_lmt/results.py
Comment (mimir-d) [inline src/pci_lmt/results.py:111]: a single variable here makes the assumption that you have a single step. consider the following application

```
r = OcptvReporter()
r.start_run()  # start run0
r.start_step()  # start step0
r.start_step()  # start step1
r.end_step()  # end step1, with NO WAY of ending step 0, which is lost; steps, as well, cannot be nested
r.end_run()  # end run0
```

the way this is avoided in ocptv api is that only the `run` objects can create steps and the steps are scoped (as implied by their reference binding)

File: src/pci_lmt/results.py
Comment (mimir-d) [inline src/pci_lmt/results.py:113]: this api design allows you to call `end_run` before `start_run`, in which case, code panics since `self._run` will be None

**Review by mimir-d — COMMENTED** [review]
the design of the `OctTvOutputReporter` thin wrapper breaks fundamental properties coming from the ocptv api lib

while this isnt mandatory to fix, i would advise to consider them since this introduces code failure modes and tech debt, which may explode later and you or another maintainer will have to address

**Comment by sksekar** [comment]
> the design of the `OctTvOutputReporter` thin wrapper breaks fundamental properties coming from the ocptv api lib

I'm trying to add the OCPTV output format natively but with very minimal changes to the main code. The thin wrapper allows me to handle OCPTV related code in a single class with very little modification to the main code.

> while this isnt mandatory to fix, i would advise to consider them since this introduces code failure modes and tech debt, which may explode later and you or another maintainer will have to address

Understood.  I would love to know if there is a way to retain the benefits of the OCPTV APIs (esp contextmanager) and still contain the implementation in it's own Reporter sub-class.

File: src/pci_lmt/results.py
Comment (sksekar) [reply src/pci_lmt/results.py:111]: Good catch. Looks like one way to avoid this is to implement the start_run/start_step as contextmanagers and invoke end_run/end_step as part of automatic cleanup. Would like to get your inputs before diving deeper.

File: src/pci_lmt/results.py
Comment (sksekar) [reply src/pci_lmt/results.py:113]: Good catch. There are ways to avoid (check for None) but that would make the code clumsy. Would like to wait for your feedback/guidance in other comments before fixing this.

File: src/pci_lmt/results.py
Comment (mimir-d) [reply src/pci_lmt/results.py:111]: yes, this is an option i highlighted above. examples https://github.com/opencomputeproject/ocp-diag-core-python/blob/dev/examples/measurements.py

**Comment by mimir-d** [comment]
i left a link with api examples above; it would be sufficient to follow that (how the api use contextmanager) in your wrapper class, since you want to keep that reporter interface. At the end of the day, try to write some code and then do the mental exercise of seeing what you can do with it. It doesnt have to be perfect, but if there's easy ways to misuse it (as i demonstrated above), then it can likely be improved

File: src/pci_lmt/results.py
Comment (sksekar) [reply src/pci_lmt/results.py]: Thanks for the suggestion and pointers. I've incorporated your comments.

File: src/pci_lmt/results.py
Comment (sksekar) [reply src/pci_lmt/results.py]: Thanks for the suggestion and pointers. I've incorporated your comments.

File: src/pci_lmt/results.py
Comment (sksekar) [reply src/pci_lmt/results.py]: Addressed this by using contextmanagers

File: src/pci_lmt/results.py
Comment (mimir-d) [inline src/pci_lmt/results.py:126]: you dont need this `self._step` anymore; removing it will avoid the initial "lost step" problem.

this func can simply be
```
@contextmanager
def start_step(self, name: str):
  step = self._run.add_step(name)
  with step.scope():
    yield
```

Comment (mimir-d) [reply src/pci_lmt/results.py:126]: similar thing in `start_run`, so you dont have to deal with the exceptions

Comment (sksekar) [reply src/pci_lmt/results.py:126]: > you dont need this self._step anymore; removing it will avoid the initial "lost step" problem.

I need to have self._step (as an object variable), since it's being used in the "write" method for adding measurement ("self._step.add_measurement"). One way to avoid is to perform "add_step" outside of the wrapper class (like it's done in the measurements.py you pointed earlier). Let me make the change.

> similar thing in start_run, so you dont have to deal with the exceptions

I'm not sure how I can avoid self._run unless I create the TestRun object outside of the wrapper class too. I prefer to have this as-is. Any concerns?

Comment (mimir-d) [reply src/pci_lmt/results.py:126]: yeah, had a feeling i was missing something

> I need to have self._step (as an object variable), since it's being used in the "write"

since you need to satisfy this `Reporter` interface, then id suggest a hard fail when trying to create a new step when the last one wasnt closed. Your api as is, with the yield, does not stop you from doing something like

```
r = Reporter()
with r.start_step("step1"):
  with r.start_step("step2"):
    pass
  r.write()  # this will incorrectly use step2
```
hence why `start_step("step2")` should fail with an exception here, at least to let the user know theyre using it incorrectly. There are ways of structurally doing this, but maybe that's too complicated at the moment.

> TestRun object outside of the wrapper class too. I prefer to have this as-is. Any concerns?

you can use the "throw invalid usage" exception pattern here too

Comment (sksekar) [reply src/pci_lmt/results.py:129]: > then id suggest a hard fail when trying to create a new step when the last one wasnt closed.

Thanks for pointing this out with a valid test-case. Handled this for TestStep by adding a check for None.

I didn't change anything for TestRun since I don't think it's needed. I believe the TestRun is per test so the chances of user having multiple instances are low. Thoughts?

Comment (mimir-d) [reply src/pci_lmt/results.py:129]: yeah, i suppose this being python, having these runtime checks should be sufficient to remove the risk of having to maintain this code unnecessarily later on.

last version is ok, if you want to merge, i'll have to look at the ocptv api types failure

Comment (sksekar) [reply src/pci_lmt/results.py:129]: Thanks for the feedback. If you are OK, i would like to merge the latest changes so that I can continue to work on validators and integrate more OCP-TV features.

Comment (mimir-d) [reply src/pci_lmt/results.py:129]: give me a day or 2 to fix the api types so we have all the tests working here

**Review by mimir-d — APPROVED** [review]

**Comment by sksekar** [comment]
Thank you so much @mimir-d. Appreciate your support (during the review) and quick turnaround (in fixing the APIs).

**Pull request merged**
